MovingBlocks/Terasology
# Crouching mode change in server/client correctly

### Contains

Begins to solve #2651

Something I identified when doing debugging was that the KineticCharacterMover(What is used to allow players to simulate across multiple clients/server) did not actually have any knowledge of crouching. I made an addition that would allow for crouching to pass, by adding a parameter to the CharacterMoveInputEvent.

This properly allows for the server and clients to know each other's actual mode and state correctly(crouching, walking, climbing, etc). But the visuals and actual representation of moving is beyond my knowledge as it is not directly tied to the mode of the player and therefore I am currently unable to solve the visual glitches.

The problem for the visual glitches can be seen if you look at the states received  from the server in the onCharacterStateReceived in ClientCharacterPredictionSystem. The current Mover doesn't correctly translate the player's position for the state based on the crouch and therefore the client is constantly translating between it's intended location(the crouched position), and the server's knowledge(in crouch, but location wise still standing). This results in the server as treating the clients as always standing visually(and then physically if they go to do any action while crouching and possibly causing desync issues).

### How to test

Addition of a few new unit tests for the new CharacterMoveInputEvent/KineticCharacterMover.

Inside of localPlayerSystem have a debug for outputting the character's movementMode, then also in ServerCharacterPredicitionSystem output the movementMode at each update for each entity. Verify the matching for both client and servers. (Let me know if more clarification is needed)

### Outstanding before merging

In summation this is the current progress

- [X] MovementModes transferred correctly between client/server
- [ ] Correct state locations transferred between client/server
- [ ] Possible refactoring of location/movement based on the movement modes and a centralized location of the character
